src/wayland/shell/xdg/handlers/surface.rs

```diff
@@ -66,7 +66,7 @@ where
                 data.known_surfaces
                     .lock()
                     .unwrap()
-                    .remove(&xdg_surface.downgrade());
+                    .shift_remove(&xdg_surface.downgrade());
```

Comment:
This does not seem to rely on the order, so `swap_remove` should be fine and has `O(1)` instead of `O(n)`.

Reply:
I was a bit confused why this uses an index map the first place and the only reasoning I could come up was, that it wants to be iterated in insertion order? Then we would break this here.

Reply:
tbh, no idea, but all it does is to insert the surface on construction, remove it on destruction and enforce that are not surfaces alive on destruction of `xdg_wm_base` which would be a protocol violation.